packages/arb-bridge-eth/contracts/arch/OneStepProof.sol

```diff
@@ -20,922 +20,772 @@ pragma solidity ^0.5.11;
+            context.startMachine.addDataStackValue(stackVals[i]);
         }
-        uint256 c;
-        assembly {
-            c := smod(a, b)
+        for (i = 0; i < auxstackCount; i++) {
+            context.startMachine.addAuxStackValue(auxstackVals[i]);
         }
-        machine.addDataStackInt(c);
-        return true;
-    }
 
-    function executeAddmodInsn(
-        Machine.Data memory machine,
-        Value.Data memory val1,
-        Value.Data memory val2,
-        Value.Data memory val3
-    ) internal pure returns (bool) {
-        if (!val1.isInt() || !val2.isInt()) {
-            return false;
-        }
-        uint256 a = val1.intVal;
-        uint256 b = val2.intVal;
-        uint256 m = val3.intVal;
-        if (m == 0) {
-            return false;
-        }
-        uint256 c;
-        assembly {
-            c := addmod(a, b, m)
-        }
-        machine.addDataStackInt(c);
-        return true;
+        return context;
     }
 
-    function executeMulmodInsn(
-        Machine.Data memory machine,
-        Value.Data memory val1,
-        Value.Data memory val2,
-        Value.Data memory val3
-    ) internal pure returns (bool) {
-        if (!val1.isInt() || !val2.isInt()) {
-            return false;
-        }
-        uint256 a = val1.intVal;
-        uint256 b = val2.intVal;
-        uint256 m = val3.intVal;
-        if (m == 0) {
-            return false;
-        }
-        uint256 c;
-        assembly {
-            c := mulmod(a, b, m)
+    function executeOp(AssertionContext memory context) internal pure {
+        (
+            uint256 dataPopCount,
+            uint256 auxPopCount,
+            uint64 gasCost,
+            function(AssertionContext memory) internal pure impl
+        ) = opInfo(context.opcode);
+        context.gas = gasCost;
+
+        // Update end machine gas remaining before running opcode
+        // No need to overflow check since the check for whether we
+        // have sufficient gas fixes the overflow case
+        context.afterMachine.arbGasRemaining =
+            context.afterMachine.arbGasRemaining -
+            gasCost;
+
+        if (context.startMachine.arbGasRemaining < gasCost) {
+            context.afterMachine.arbGasRemaining = MAX_UINT256;
+            context.handleError();
+            return;
         }
-        machine.addDataStackInt(c);
-        return true;
-    }
 
-    function executeExpInsn(
-        Machine.Data memory machine,
-        Value.Data memory val1,
-        Value.Data memory val2
-    ) internal pure returns (bool) {
-        if (!val1.isInt() || !val2.isInt()) {
-            return false;
+        if (context.stack.length < dataPopCount) {
+            // If we have insufficient values, reject the proof unless the stack has been fully exhausted
+            require(
+                context.afterMachine.dataStack.hash() ==
+                    Value.newEmptyTuple().hash(),
+                "stack not empty"
```

Comment:
This is not a super-helpful error message. Something like "stack item missing from proof" seems better.

Reply:
Fixed with your suggested message